### PR TITLE
relay: support relay log in v2.0

### LINKED ISSUE #1234: Bring relay support back in v2.0
## Feature Request

**Is your feature request related to a problem? Please describe:**

in v1.0, DM supports relay log. in current 2.0, it's unsupported yet.

so, bring relay support back in v2.0.

**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve? 
part of #1234 

### What is changed and how it works?
- [x]  support TLS in relay
- [x]  support gtid in relay

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

### CODE REVIEW

File: dm/worker/subtask.go [hunk 1]
```diff
@@ -600,10 +602,24 @@ func (st *SubTask) unitTransWaitCondition(subTaskCtx context.Context) error {
...
+			if st.cfg.EnableGTID {
+				gset1, err := gtid.ParserGTID(st.cfg.Flavor, loadStatus.MetaBinlogGTID)
+				if err != nil {
+					st.l.Error("fail to parse gtid", log.ShortError(err))
+				}
+				gset2, err := gtid.ParserGTID(st.cfg.Flavor, relayStatus.RelayBinlogGtid)
+				if err != nil {
+					st.l.Error("fail to parse gtid", log.ShortError(err))
+				}
+				rc, ok := binlog.CompareGTID(gset1, gset2)
```
Comment (lance6716) [inline dm/worker/subtask.go:614]: maybe we should handle not `ok` case instead of dead loop

File: pkg/binlog/event/event.go [hunk 2]
```diff
@@ -201,7 +201,7 @@ func GenRotateEvent(header *replication.EventHeader, latestPos uint32, nextLogNa
 //   a. https://github.com/vitessio/vitess/blob/28e7e5503a6c3d3b18d4925d95f23ebcb6f25c8e/go/mysql/binlog_event_mysql56.go#L56
 //   b. https://dev.mysql.com/doc/internals/en/com-binlog-dump-gtid.html
 func GenPreviousGTIDsEvent(header *replication.EventHeader, latestPos uint32, gSet gtid.Set) (*replication.BinlogEvent, error) {
-	if gSet == nil || len(gSet.String()) == 0 {
```
Comment (lance6716) [inline pkg/binlog/event/event.go:204]: not understand why remove this 🤔

**Review by lance6716 — COMMENTED** [review]
will review later

File: dm/worker/subtask.go (hunk 1 above)
Comment (GMHDBJD) [reply dm/worker/subtask.go:614]: If not ok, will compare pos below(L619).

Comment (GMHDBJD) [reply dm/worker/subtask.go:614]: Generate nil previous gtid event. Like `reset master`

File: pkg/binlog/event/event.go (hunk 2 above)
Comment (GMHDBJD) [reply pkg/binlog/event/event.go:204]: Generate nil previous gtid event. Like reset master.

File: pkg/streamer/reader.go [hunk 3]
```diff
@@ -103,9 +117,203 @@ func (r *BinlogReader) checkRelayPos(pos mysql.Position) error {
...
+			// if PreviousGITDsEvent contain the gset, go to previous file
+			if gs.Contain(gset) {
+				return 0, nil
+			}
+		case *replication.RotateEvent:
+			if e.Header.Timestamp != 0 && e.Header.LogPos != 0 {
+				// gset is the last txn in file
+				return lastPos, nil
+			}
+		case *replication.XIDEvent:
```
Comment (lance6716) [inline pkg/streamer/reader.go:246]: `		case *replication.XIDEvent, *replication.QueryEvent:`

File: pkg/streamer/reader.go [hunk 4]
```diff
@@ -103,9 +117,203 @@ func (r *BinlogReader) checkRelayPos(pos mysql.Position) error {
...
+			offset = int64(e.Header.LogPos)
+
+			select {
+			case s.ch <- e:
+			case <-ctx2.Done():
+			}
+			return nil
+		}
+
+		err := parser.ParseFile(filePath, offset, onEventFunc)
```
Comment (lance6716) [inline pkg/streamer/reader.go:194]: seems this `ParseFile` will open file, read some content and parse, seek, and close. maybe we could use `ParseSingleEvent` or something to reuse last parser internal status

File: pkg/streamer/reader.go [hunk 5]
```diff
@@ -103,9 +117,203 @@ func (r *BinlogReader) checkRelayPos(pos mysql.Position) error {
...
+		return nil, err
+	}
+
+	uuidDir := path.Join(r.cfg.RelayDir, uuid)
+	allFiles, err := CollectAllBinlogFiles(uuidDir)
+	if err != nil {
+		return nil, err
+	}
+
+	// iterate files from the newest one
```
Comment (lance6716) [inline pkg/streamer/reader.go:292]: in future we could use a binary search to find "the last file returns non-zero `pos`" and use that `pos`

File: pkg/streamer/reader.go [hunk 6]
```diff
@@ -314,10 +563,38 @@ func (r *BinlogReader) parseFile(
...
+				latestPos = int64(e.Header.LogPos)
+				break
+			}
+			GTID := ev.GTID
+			err2 := r.advanceCurrentGtidSet(fmt.Sprintf("%d-%d-%d", GTID.DomainID, GTID.ServerID, GTID.SequenceNumber))
+			if err2 != nil {
+				return errors.Trace(err2)
+			}
+			latestPos = int64(e.Header.LogPos)
+		case *replication.XIDEvent:
```
Comment (lance6716) [inline pkg/streamer/reader.go:592]: ditto

File: dm/worker/subtask.go (hunk 1 above)
Comment (lance6716) [reply dm/worker/subtask.go:614]: I didn't learn GTID much, will there be not comparable `gset1` `gset2`, and is falling to binlog position compare a good way to solve?

**Review by lance6716 — COMMENTED** [review]
will look test later

File: dm/worker/subtask.go (hunk 1 above)
Comment (GMHDBJD) [reply dm/worker/subtask.go:614]: If gset1 can't compare gset2, there must be some error occured and syncer also can't sync from dump_data/metadata. I think both deed loop or report a error is ok?

File: pkg/streamer/reader.go (hunk 4 above)
Comment (GMHDBJD) [reply pkg/streamer/reader.go:194]: We need to parse the whole file until we meet first previous_gtid_event/xid_event/query_event which not contain gset. If use ParseSingleEvent, we still need these logic(seek/close...)

File: pkg/streamer/reader.go (hunk 6 above)
Comment (GMHDBJD) [reply pkg/streamer/reader.go:590]: below use ev.Gset, so cannot combine them together.

File: pkg/streamer/reader.go (hunk 4 above)
Comment (lance6716) [reply pkg/streamer/reader.go:194]: I wrongly read this file, thought it's in a for-loop 😢

File: pkg/streamer/reader_test.go [hunk 7]
```diff
@@ -604,13 +624,210 @@ func (t *testReaderSuite) TestStartSync(c *C) {
...
+				},
+			},
+		},
+	}
+
+	for _, subDir := range testCase {
+		r.uuids = append(r.uuids, subDir.uuid)
+	}
+
+	// wirte index file
```
Comment (lance6716) [inline pkg/streamer/reader_test.go:740]:
```suggestion
	// write index file
```

**Review by lance6716 — COMMENTED** [review]
rest LGTM

File: pkg/streamer/reader.go [hunk 8]
```diff
@@ -103,9 +117,206 @@ func (r *BinlogReader) checkRelayPos(pos mysql.Position) error {
...
+			if err != nil {
+				return 0, err
+			}
+			// if PreviousGITDsEvent contain the gset, go to previous file
+			if gs.Contain(gset) {
+				// if it's the oldest file and gs equals gset, continue
+				if oldest && gset.Contain(gs) {
+					lastPos = e.Header.LogPos
+					continue
+				}
```
Comment (GMHDBJD) [inline pkg/streamer/reader.go:243]: mysql.000001: [..., gtid, query, heartbeat, ..., rotate]
mysql.000002: [format, previous_gtid]
We cant start from mysql.000001's heartbeat event or mysql.000002's previous_gtid event.
What are your opinions? @lance6716 @csuzhangxc

**Comment by lance6716** [comment]
LGTM

**Comment by csuzhangxc** [comment]
@GMHDBJD please resolve conflicts, 😹

File: tests/shardddl3/conf/source1.yaml [hunk 9]
```diff
@@ -1,7 +1,7 @@
 source-id: mysql-replica-01
 flavor: ''
 enable-gtid: false
-enable-relay: false
+enable-relay: true
```
Comment (csuzhangxc) [inline tests/shardddl3/conf/source1.yaml:4]: How about enabling GTID only for one source? (and keep another one use remote binlog).

File: dm/worker/subtask.go [hunk 10]
```diff
@@ -612,10 +614,24 @@ func (st *SubTask) unitTransWaitCondition(subTaskCtx context.Context) error {
 			if err != nil {
 				return terror.WithClass(err, terror.ClassDMWorker)
 			}
+			if st.cfg.EnableGTID {
+				gset1, err := gtid.ParserGTID(st.cfg.Flavor, loadStatus.MetaBinlogGTID)
+				if err != nil {
```
Comment (csuzhangxc) [inline dm/worker/subtask.go:619]: `if err != nil {`, we still do `rc, ok := binlog.CompareGTID(gset1, gset2)`, is there any problem?

File: dm/worker/subtask.go [hunk 11]
```diff
@@ -612,10 +614,24 @@ func (st *SubTask) unitTransWaitCondition(subTaskCtx context.Context) error {
 			if err != nil {
 				return terror.WithClass(err, terror.ClassDMWorker)
 			}
+			if st.cfg.EnableGTID {
```
Comment (csuzhangxc) [inline dm/worker/subtask.go:617]: we can put this code block before `pos2, err := utils.DecodeBinlogPosition(relayStatus.RelayBinlog)`.

File: dm/worker/subtask.go (hunk 1 above)
Comment (csuzhangxc) [reply dm/worker/subtask.go:626]: I prefer to return the error explicitly and fast failure.

File: pkg/streamer/reader.go [hunk 12]
```diff
@@ -215,14 +175,17 @@ func (r *BinlogReader) GetFilePosByGTID(ctx context.Context, filePath string, gs
 		default:
 		}
 
-		e, err := s.GetEvent(ctx)
+		ctx2, cancel := context.WithTimeout(ctx, time.Second)
+		e, err := fileReader.GetEvent(ctx2)
+		cancel()
 		if err != nil {
+			// reach end of file
+			if errors.Cause(err) == context.DeadlineExceeded {
```
Comment (csuzhangxc) [inline pkg/streamer/reader.go:182]: Is there any problem if the parent `ctx` deadline exceeded? How about checking `lastPos` with the file size? (may be still there are problems if the file is being written?

Comment (GMHDBJD) [reply pkg/streamer/reader.go:182]: use context.background() as parent ctx of ctx2?

Comment (csuzhangxc) [reply pkg/streamer/reader.go:182]: if use `context.background()` as the parent, can the user to cancel the process?

**Review by csuzhangxc — APPROVED** [review]
LGTM

**Pull request merged**